routing/cross_mwm_road_graph.cpp

```diff
@@ -135,34 +135,35 @@ BorderCross CrossMwmGraph::ConstructBorderCross(OutgoingCrossNode const & startN
   if (it != m_cachedNextNodes.end())
     return it->second;
 
-  // Cache miss case.
-  BorderCross cross;
-  if (!ConstructBorderCrossImpl(startNode, currentMapping, cross))
-    return BorderCross();
-  m_cachedNextNodes.insert(make_pair(key, cross));
-  return cross;
+  // Cache miss case. I suppose that cache will be invalidated after a search.
+  vector<BorderCross> crosses;
+  ConstructBorderCrossImpl(startNode, currentMapping, crosses);
+  auto const inserted = m_cachedNextNodes.insert(make_pair(key, crosses));
+  return inserted.first->second;
 }
 
 bool CrossMwmGraph::ConstructBorderCrossImpl(OutgoingCrossNode const & startNode,
                                              TRoutingMappingPtr const & currentMapping,
-                                             BorderCross & cross) const
+                                             vector<BorderCross> & crosses) const
 {
+  auto const fromCross = CrossNode(startNode.m_nodeId, currentMapping->GetMwmId(), startNode.m_point);
   string const & nextMwm = currentMapping->m_crossContext.GetOutgoingMwmName(startNode);
   TRoutingMappingPtr nextMapping = m_indexManager.GetMappingByName(nextMwm);
   // If we haven't this routing file, we skip this path.
   if (!nextMapping->IsValid())
     return false;
+  crosses.clear();
```

Comment:
Yes, it is redundant. But I will leave the ASSERT expression to check if crosses are empty.